### PR TITLE
Add missing knowledge checks and learning outcomes to ruby/object-oriented-programming lesson

### LINKED ISSUE #22922: Add missing Knowledge Checks, Learning Outcomes, and Knowledge Check links to Ruby/Object Oriented Programing
Related to #22467

This task is to go through the lesson referenced in the title of this issue and (as applicable):

- Add missing Learning Outcomes
- Add missing Knowledge Checks
- Add answer links in the Knowledge Checks

If the lesson already has Learning Outcomes or Knowledge Checks they do not need to be added to. Many lessons only need answer links added.

Knowledge check links should consist of an `<a>` tag which adds the `knowledge-check-link` class and links to the part of the lesson (or the external resource that is linked in the lesson) where the answer to the Knowledge Check can be found. Knowledge Check links should **not** point to external resources that weren't linked to as part of the lesson content or assignment. You can link to a specific line in the lesson like so:

```html
Lorem ipsum dolor sit amet, <span id='example-knowledge-check'>consectetur adipiscing elit</span>, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua

...
<!-- This a tag just goes to the page where the span with this id lives -->
<a class='knowledge-check-link' href='#example-knowledge-check'>How do you return a value from a function?</a>
```
Or you can link to just a section of the lesson by just linking to the section heading and following [these](https://github.com/TheOdinProject/curriculum/wiki/Contributing-to-Wiki#anchor-linking) rules for determining how to format the link. Example:

```html
### Lorem ipsum dolor sit amet

sed do eiusmod tempor incididunt ut labore et dolore magna aliqua

...
<!-- This a tag just goes to a heading on the page-->
<a class='knowledge-check-link' href='#lorem-ipsum-dolor-sit-amet'>How do you return a value from a function?</a>
```

### Acceptance Criteria

- [ ] Lesson referenced in the title has Learning Outcomes, Knowledge Checks, and Knowledge Check links
- [ ] Knowledge Check links all link to the proper place in the lesson, or the proper external resource that was linked to earlier in the lesson
- [ ] Knowledge Check styling should **not** change

Please do not start work on this unless you've been assigned to it. Comment on this issue and ping me @wise-king-sullyman if you would like to be assigned to it.

Additionally, we would like for contributors to only take on one of these issues at a time. Please only request assignment for (and begin work on) an issue after the Learning Outcome/Knowledge Check/Knowledge Check link issue you're currently assigned to has been approved and merged.

Please check out [this new post on our wiki](https://github.com/TheOdinProject/curriculum/wiki/Learning-Outcomes-vs-Knowledge-Checks) about Learning Outcomes and Knowledge Checks if you need to write new Learning Outcomes or Knowledge Checks for this issue.

### PR DESCRIPTION
**Pull request by LeoKirasic**
- [x] You have read [The Odin Projects Contributing Guide](https://github.com/TheOdinProject/curriculum/blob/main/CONTRIBUTING.md).
 - [x] The PR title summarizes the change and where it happened, for example: "Fixes punctuation in Clean Code lesson".
 
#### 1.Describe the changes made and include why they are necessary or important:

I have added knowledge checks and proper learning outcomes to the ruby/object-oriented-programming lesson. 

#### 2. If this PR is related to an open issue please reference it with the `#` operator and the issue number (and any [relevant keywords](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests)) below:

closes #22922

### CODE REVIEW

File: ruby_programming/object_oriented_programming_basics/lesson_oop.md
Comment (wise-king-sullyman) [inline ruby_programming/object_oriented_programming_basics/lesson_oop.md:68]: Are these questions directly answered in the linked resource? I'm not seeing them but I 100% could just be missing them.

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
This looks pretty solid! Just have a question on a couple of your Knowledge Checks.

I've requested the opinions of others regarding all the questions being removed from the current Learning Outcomes. I agree with your assessment that they don't belong there, but I want to see what the rest of the team thinks about them just removed vs if we want to try and move them somewhere else.

**Comment by LeoKirasic** [comment]
> This looks pretty solid! Just have a question on a couple of your Knowledge Checks.
> 
> I've requested the opinions of others regarding all the questions being removed from the current Learning Outcomes. I agree with your assessment that they don't belong there, but I want to see what the rest of the team thinks about them just removed vs if we want to try and move them somewhere else.

I don't think that all of them are answered in the lesson which becomes an issue when moving them to knowledge checks under the new requirements.

File: ruby_programming/object_oriented_programming_basics/lesson_oop.md
Comment (LeoKirasic) [reply ruby_programming/object_oriented_programming_basics/lesson_oop.md:68]: Not directly answered. Eriks post indirectly answers it. I'll update the links to a more directly answered resource.

**Review by wise-king-sullyman — APPROVED** [review]
This looks solid to me, just still waiting on some feedback from other maintainers so I won't be merging it quite yet.

**Review by rlmoser99 — CHANGES_REQUESTED** [review]
This is a really big lesson and I like the idea of slimming down the large amount of Learning Outcomes. However, I would really like to see a question about use of instance variables in the Knowledge Check because we see people using global and class variables when they should be using instance variables. Possibly even a question about why they should not use global or class variables.

**Comment by rlmoser99** [comment]
@LeoKirasic Did you have some changes that you wanted me to look at? Or do you need more direction on my feedback?

**Comment by LeoKirasic** [comment]
@rlmoser99 I accidentally didn't push the commit, I apologise for that. 
I have added the changes you requested.

File: ruby_programming/object_oriented_programming_basics/lesson_oop.md
Comment (rlmoser99) [inline ruby_programming/object_oriented_programming_basics/lesson_oop.md:73]: Thanks for looking for where these two questions would be answered in this LONG lesson! After reviewing these resources, I do not think my proposed "why they should not use global or class variables" is really explained (no wonder people get confused here). So I've adjusting the wording to what I think that resource actually does explain. 

In addition, as a nit pick, I suggested a change for the instance variable one, simply because I do like like the shift of wording here to using "we" (in general I don't like the "we" wording, like "How are we doing today?" but I realize that I am could be in the minority on this, so this is just simply a suggestion.

```suggestion
- <a class="knowledge-check-link" href="https://launchschool.com/books/oo_ruby/read/classes_and_objects_part1#instancevariables">What is the purpose of instance variables?</a>
- <a class="knowledge-check-link" href="https://www.eriktrautman.com/posts/ruby-explained-classes">What is the difference between class variables and instance variables?</a>
```

**Review by rlmoser99 — CHANGES_REQUESTED** [review]
Thanks for taking the time to update this, just some small wording suggestions.

**Review by rlmoser99 — APPROVED** [review]
Thanks for taking the time to make these changes! 👍

**Pull request merged**
